entity-framework/core/testing/index.md

```diff
@@ -1,117 +1,116 @@
+3. Isolation is indeed a hurdle when running tests against a real database, as tests may modify data and interfere with one another. However, there are various techniques to provide isolation in database testing scenarios; we concentrate on these in the [integration testing docs](xref:core/testing/integration-testing)).
 
-However, having every developer run tests against SQL Azure while actively working on the code would be both slow and expensive.
-This illustrates the main trade-off involved throughout these approaches: when is it appropriate to deviate from the production database system so as to improve test efficiency?
+The above is not meant to disparage unit testing or to argue against writing them. For one thing, unit tests are necessary for some scenarios which cannot be tested with integration tests, such as simulating database failure. However, in our experience, users frequently shy away from integration testing for the above reasons, believing it's slow, hard or unreliable, when that isn't necessarily the case. [The dedicated page on integration testing](xref:core/testing/integration-testing) aims to address this, providing guidelines and samples for writing fast, isolated database integration tests.
 
-Luckily, in this case the answer is quite easy: use local or on-premises SQL Server for developer testing.
-SQL Azure and SQL Server are extremely similar, so testing against SQL Server is usually a reasonable trade-off.
-That being said, it is still wise to run tests against SQL Azure itself before going into production.
+## Unit testing
 
-### LocalDB
+Unlike integration tests, unit tests never involve the actual database system used in production, but rather swap it with a [test double](https://en.wikipedia.org/wiki/Test_double). Below are some common methods for unit testing EF Core applications with test doubles:
 
-All the major database systems have some form of "Developer Edition" for local testing.
-SQL Server also has a feature called [LocalDB](/sql/database-engine/configure-windows/sql-server-express-localdb).
-The primary advantage of LocalDB is that it spins up the database instance on demand.
-This avoids having a database service running on your machine even when you're not running tests.
+1. Use SQLite (in-memory mode) as a database fake, replacing your production database system.
+2. Use the EF Core InMemory provider as a database fake, replacing your production database system.
+3. Mocking or stubbing `DbContext` and `DbSet`.
+4. Introduce a repository layer between EF Core and your application code, and mocking or stubbing that layer.
 
-LocalDB is not without its issues:
+Below, we'll explore what each method means, and compare it with the others. We recommend reading through the different methods to gain a full understanding, but if you're looking for a quick suggestion, we highly recommend the 4th method: introducing a repository layer.
```

Comment:
Do we highly recommend using a repository? I don't. I highly recommend not writing the kind of unit tests that require a repository.

Reply:
Made consistent and corrected.